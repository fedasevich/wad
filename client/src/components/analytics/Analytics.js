import { endOfMonth, startOfMonth } from 'date-fns';
import { observer } from 'mobx-react-lite';
import React, { useContext, useEffect, useState } from 'react';
import { Col, Row } from 'react-bootstrap';
import { Context } from '../..';
import PageProvider from '../../pages/PageProvider';
import MenuProvider from '../MenuProvider';
import DatePickerProvider from '../date-picker/DatePickerProvider';
import { AnalyticsSelect } from './AnalyticsSelect';
import './AnalyticsStyle.css';
import { AnalyticsChart } from './chart/AnalyticsChart';
import { AnalyticsPercentage } from './percentage/AnalyticsPercentage';
import { AnalyticsStatistics } from './statistics/AnalyticsStatistics';

const AnalyticsHOC = (WrappedComponent) => {
    return function (props) {
        if (props.transactions.length) {
            return <WrappedComponent {...props} />;
        } else {
            return <h2>There is no statistics</h2>;
        }
    };
};


export const Analytics = observer(() => {
    const { category } = useContext(Context)

    const [loading, setLoading] = useState(true)
    const [dateRange, setDateRange] = useState([
        {
            startDate: startOfMonth(new Date()),
            endDate: endOfMonth(new Date()),
            key: 'selection',
            action: 'month'
        }
    ]);
    const [chartRange, setChartRange] = useState('')
    const [transactions, setTransactions] = useState([])


    useEffect(() => {
        try {
            category.fetchCategoryPeriod(dateRange).then((data) => {
                setTransactions(data.rows);
                category.parseCategories(data)
            }).finally(() => setLoading(false))
        } catch (e) {
            alert(e.response.data.message);
        }
    }, [category, dateRange])

    if (loading || !category.categories.length) {
        return (<h2>loading</h2>)
    }

    const HOCAnalyticsPercentage = AnalyticsHOC(AnalyticsPercentage);
    const HOCAnalyticsStatistics = AnalyticsHOC(AnalyticsStatistics);

    return (
        <Row>
            <PageProvider.Header pageName={'Analytics'}>
                <Col xs={{ span: 5, offset: 2 }} className="d-flex align-items-center justify-content-center flex-column">
                    <DatePickerProvider dateRange={dateRange} setDateRange={setDateRange} />
                    <AnalyticsSelect chartRange={chartRange} setChartRange={setChartRange} />
                </Col>
            </PageProvider.Header >
            <Col md={{ offset: 1, span: 10 }} >
                <MenuProvider>
                    <MenuProvider.Container >
                        <AnalyticsChart transactions={transactions} chartRange={chartRange} />
                    </MenuProvider.Container>
                </MenuProvider>
            </Col>
            <Col md={{ offset: 1, span: 6 }} >
                <MenuProvider>
                    <MenuProvider.Container>
                        <HOCAnalyticsPercentage transactions={transactions} />
                    </MenuProvider.Container>
                </MenuProvider>
            </Col>
            <Col md={{ span: 4 }} >
                <MenuProvider>
                    <MenuProvider.Container>
                        <HOCAnalyticsStatistics transactions={transactions} chartRange={chartRange} />
                    </MenuProvider.Container>
                </MenuProvider>
            </Col>
        </Row>
    );
})