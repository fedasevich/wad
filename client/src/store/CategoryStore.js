import { makeAutoObservable, runInAction } from "mobx";
import { changeCategory, createCategory, deleteCategory, fetchCategory, fetchCategoryPeriod } from "../http/categoryApi";

import { clearPersistedStore, makePersistable, pausePersisting, startPersisting } from "mobx-persist-store";
import { changeTransaction, createTransaction, deleteTransaction } from "../http/transactionApi";


export default class CategoryStore {
  constructor(rootStore) {
    this.rootStore = rootStore
    this._categories = []
    this._parsedCategories = []
    this._transactions = []
    this._transactionsLimit = 5
    this._transactionsPage = 1
    this._transactionsSort = "DESC"
    this._selectedCategory = {}
    this._selectedTransaction = {}
    makeAutoObservable(this)
    makePersistable(this, {
      name: "transaction",
      properties: ['_transactionsSort', '_transactionsLimit', '_transactionsPage'],
      storage: window.localStorage
    }, { fireImmediately: true });
  }

  setCategories(categories) {
    this._categories = categories
  }

  setParsedCategories(parsedCategories) {
    this._parsedCategories = parsedCategories
  }

  setTransactions(transactions) {
    this._transactions = transactions
  }

  setSelectedCategory(category) {
    this._selectedCategory = category
  }

  setSelectedTransaction(transaction) {
    this._selectedTransaction = transaction
  }

  setTransactionsLimit(limit) {
    this._transactionsLimit = limit
  }

  setTransactionsPage(page) {
    this._transactionsPage = page
  }

  setTransactionsSort(sort) {
    this._transactionsSort = sort
  }


  clearPersistence() {
    clearPersistedStore(this)
  }

  modifyTransactionsFilter({ limit = this.transactionsLimit, page = this.transactionsPage, sort = this.transactionsSort }) {
    if (this.rootStore.userSettings.persistTransactionFilter) {
      startPersisting(this)
    } else {
      pausePersisting(this)
    }

    this.setTransactionsLimit(limit);
    this.setTransactionsPage(page);
    this.setTransactionsSort(sort);
  }

  get categories() {
    return this._categories
  }

  get parsedCategories() {
    return this._parsedCategories
  }

  get selectedTransaction() {
    return this._selectedTransaction
  }
  get transactions() {
    return this._transactions
  }

  get selectedCategory() {
    return this._selectedCategory
  }

  get transactionsPage() {
    return this._transactionsPage
  }

  get transactionsLimit() {
    return this._transactionsLimit
  }

  get transactionsSort() {
    return this._transactionsSort
  }

  createTransaction(currentOperand, selectedCategory, selectedWallet, description) {
    try {
      createTransaction(selectedCategory.id, selectedWallet.id, description ? description : selectedCategory.name, parseFloat(currentOperand))
        .then(data => {
          runInAction(() => {
            this.transactions.unshift(data)
            selectedWallet.balance -= parseFloat(currentOperand)
            this.getCategoryById(selectedCategory.id).spent += parseFloat(currentOperand)
          })
        })
    } catch (e) {
      alert(e.response.data.message);
    }
  }

  deleteTransaction(transactionId, categoryId, walletId) {
    try {
      deleteTransaction(transactionId)
        .then(() => {
          runInAction(() => {
            const transaction = this.getTransactionById(transactionId)
            const category = this.getCategoryById(categoryId) || this.getRegularCategoryById(categoryId)
            if (walletId !== -1) {
              const foundWallet = this.rootStore.wallet.getWalletById(walletId)
              foundWallet.balance += parseFloat(transaction.sum)
            }
            category.spent -= parseFloat(transaction.sum)
            this.transactions.splice(this.transactions.indexOf(transaction), 1)
          })
        })
    } catch (e) {
      alert(e.response.data.message);
    }
  }

  getTransactionById = (id) => {
    return this.transactions.find((transaction) => transaction.id === id)
  }


  getRegularCategoryById = (id) => {
    if (id === -1) return this.categories[0]
    return this.categories.find((category) => category.id === id)
  }

  getCategoryById = (id) => {
    if (id === -1) return this.parsedCategories[0]
    return this.parsedCategories.find((category) => category.id === id)
  }

  getIconIdFromCategoryById = (id) => {
    const findCategory = this.getRegularCategoryById(id)
    return findCategory?.iconId
  }

  createCategory(newCategoryName, newCategorySelectedIcon) {
    try {
      return createCategory(newCategoryName, newCategorySelectedIcon.id)
        .then(data => {
          runInAction(() => {
            this.parsedCategories.push(data)

          })
        })
    } catch (e) {
      alert(e.response.data.message);
    }
  }

  changeTransaction(id, newSum, newDescription) {
    if (!newSum && !newDescription) {
      return alert(`Not enough data`)
    }
    const transaction = this.getTransactionById(id)
    try {
      changeTransaction(transaction.id, parseFloat(newSum), newDescription)
        .then(() => {
          runInAction(() => {
            const categoryToChange = this.getRegularCategoryById(transaction.categoryId)
            const walletToChange = this.rootStore.wallet.getWalletById(transaction.walletId)
            if (newSum) {
              if (transaction.walletId !== -1) {
                walletToChange.balance = (parseFloat(walletToChange.balance) + parseFloat(transaction.sum)) - parseFloat(newSum)
              }
              if (categoryToChange.id !== -1) {
                categoryToChange.spent = (parseFloat(categoryToChange.spent) - parseFloat(transaction.sum)) + parseFloat(newSum)
              }
              transaction.sum = newSum
            }
            if (newDescription) {
              categoryToChange.description = newDescription
              transaction.description = newDescription
            }
          })
        })
    } catch (e) {
      alert(e.response.data.message);
    }
  }

  changeCategory(id, editCategory) {
    if ((editCategory || !editCategory.spent) && !editCategory.name && !Object.keys(editCategory.icon).length) {
      return alert(`Not enough data`)
    }
    try {
      return changeCategory(id, parseFloat(editCategory.spent), editCategory.name, editCategory.icon.id)
        .then((data) => {
          runInAction(() => {
            const categoryToEdit = this.getCategoryById(id)
            Object.assign(categoryToEdit, data)
          })
        })
    } catch (e) {
      alert(e.response.data.message);
    }
  }

  fetchCategoryPeriod(dateRange) {
    if (!dateRange || !dateRange[0].startDate || !dateRange[0].endDate) return;

    const startDate = dateRange[0].startDate.toISOString();
    const endDate = dateRange[0].endDate.toISOString();

    return fetchCategoryPeriod(startDate, endDate)
      .then(data => {
        return data
      })
      .catch(error => {
        alert(error.message);
      });
  }

  parseCategories(data) {
    runInAction(() => {
      this._parsedCategories = this._categories.map(category => {
        const { id } = category;
        const spent = data.rows.reduce((total, row) => {
          if (row.categoryId === id) {
            return total + parseFloat(row.sum);
          }
          return total;
        }, 0);

        return {
          ...category,
          spent,
        };
      });
    });
  }

  fetchCategory() {
    try {
      fetchCategory().then(data => {
        runInAction(() => {
          this.setCategories(data)
        })
      })
    } catch (e) {
      alert(e.response.data.message);
    }
  }

  async deleteCategory(id) {
    try {
      await deleteCategory(id)
        .then(() => {
          runInAction(() => {
            const categoryIndex = this._parsedCategories.findIndex(category => category.id === id)
            this._parsedCategories.splice(categoryIndex, 1)
          })
        })
    } catch (e) {
      alert(e.response.data.message);
    }
  }

}