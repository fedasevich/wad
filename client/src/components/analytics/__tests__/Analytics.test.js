import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import { mockTransactions } from '../../../../setupTests';
import CategoryStore from '../../../store/CategoryStore';
import { Analytics } from '../Analytics';

jest.mock('../../../store/CategoryStore.js', () => ({
  fetchCategoryPeriod: jest.fn()
}));

jest.mock('react-chartjs-2', () => ({
  Bar: () => <canvas />
}));

jest.mock('../../../store');

describe('Analytics.js', () => {
  beforeAll(() => {
    jest.useFakeTimers('modern');
    jest.setSystemTime(new Date(2023, 7, 30));
  });

  beforeEach(async () => {
    CategoryStore.fetchCategoryPeriod.mockResolvedValue({ rows: mockTransactions });
  });

  it('matches snapshot', async () => {
    const { container } = await act(async () => {
      return render(<Analytics />);
    });
    expect(container).toMatchSnapshot();
  });

  afterAll(() => {
    jest.useRealTimers();
  });
});
