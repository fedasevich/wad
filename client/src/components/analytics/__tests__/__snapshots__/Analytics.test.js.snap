// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Analytics.js matches snapshot 1`] = `
<div>
  <div
    class="row"
  >
    <header>
      <div
        class="row"
      >
        <div
          class="col-4 offset-sm-1"
        >
          <h1
            class="mt-5 mb-5 fw-bold"
          >
            Analytics
          </h1>
        </div>
        <div
          class="d-flex align-items-center justify-content-center flex-column col-5 offset-2"
        >
          <div
            class="undefined d-flex fw-medium justify-content-evenly cursor-pointer align-items-center"
          >
            <button
              class="px-1"
              type="button"
            >
              <svg>
                CaretLeft.svg
              </svg>
            </button>
            <button
              class="px-1"
              type="button"
            >
              <h3>
                August 2023
              </h3>
            </button>
            <button
              class="px-1"
              type="button"
            >
              <svg>
                CaretRight.svg
              </svg>
            </button>
          </div>
          <div
            class="undefined d-flex fw-medium justify-content-evenly cursor-pointer align-items-center"
          >
            <button
              class="px-1"
              type="button"
            >
              <svg>
                CaretLeft.svg
              </svg>
            </button>
            <button
              class="px-1"
              type="button"
            >
              <h3>
                week
              </h3>
            </button>
            <button
              class="px-1"
              type="button"
            >
              <svg>
                CaretRight.svg
              </svg>
            </button>
          </div>
        </div>
      </div>
    </header>
    <div
      class="col-md-10 offset-md-1"
    >
      <div
        class=" menu mb-3"
      >
        <div
          class="menu-container d-flex flex-column "
        >
          <div
            class="analytics-chart-container"
          >
            <div
              class="analytics-chart "
              data-testid="analytics-chart"
            >
              <canvas />
            </div>
            <div
              class="d-flex d-md-none justify-content-end expand-button"
            >
              <button
                class="bg-light-blue border-0 btn btn-primary"
                data-testid="expand-button"
                type="button"
              >
                <svg>
                  CaretDown.svg
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="percentage col-md-6 offset-md-1"
    >
      <div
        class="percentage-menu menu mb-3"
      >
        <div
          class="menu-container d-flex flex-column "
        >
          <h3
            class="fw-medium"
          >
            Percentage
          </h3>
          <button
            class="mt-2 d-flex flex-row align-items-center justify-content-center"
            data-testid="percentage-item"
            disabled=""
            type="button"
          >
            <div
              class=" position-relative categoryIcon me-3"
            >
              <div
                class="bg-main-blue component-one-third-border-radius"
                style="background-color: rgb(122, 175, 252);"
              >
                <svg
                  fill="none"
                  height="50"
                  viewBox="0 0 43 43"
                  width="50"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M8.7275 16.0725C7.83123 15.0833 7.34967 13.7874 7.38253 12.453C7.41538 11.1185 7.96013 9.84786 8.904 8.904C9.84786 7.96013 11.1185 7.41538 12.453 7.38253C13.7874 7.34967 15.0833 7.83123 16.0725 8.7275L34.2725 26.9275C35.1688 27.9167 35.6503 29.2126 35.6175 30.547C35.5846 31.8815 35.0399 33.1521 34.096 34.096C33.1521 35.0399 31.8815 35.5846 30.547 35.6175C29.2126 35.6503 27.9167 35.1688 26.9275 34.2725L8.7275 16.0725Z"
                    stroke="white"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                  />
                  <path
                    d="M26.9275 8.7275C27.9167 7.83123 29.2126 7.34967 30.547 7.38253C31.8815 7.41538 33.1521 7.96013 34.096 8.904C35.0399 9.84786 35.5846 11.1185 35.6175 12.453C35.6503 13.7874 35.1688 15.0833 34.2725 16.0725L16.0725 34.2725C15.0833 35.1688 13.7874 35.6503 12.453 35.6175C11.1185 35.5846 9.84786 35.0399 8.904 34.096C7.96013 33.1521 7.41538 31.8815 7.38253 30.547C7.34967 29.2126 7.83123 27.9167 8.7275 26.9275L26.9275 8.7275Z"
                    stroke="white"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                  />
                  <path
                    d="M21.5 22.15C21.859 22.15 22.15 21.859 22.15 21.5C22.15 21.141 21.859 20.85 21.5 20.85C21.141 20.85 20.85 21.141 20.85 21.5C20.85 21.859 21.141 22.15 21.5 22.15Z"
                    opacity="0.5"
                    stroke="white"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                  />
                </svg>
              </div>
            </div>
            <div
              class="d-flex flex-column w-100"
            >
              <div
                class="d-flex justify-content-between"
              >
                <p
                  class="fs-4 mb-0"
                >
                  ggg
                </p>
                <p
                  class="fs-4 mb-0 fw-medium"
                >
                  1700
                </p>
              </div>
              <div
                class="d-flex w-100 justify-content-between"
              >
                <div
                  class="progress w-100 mt-1 me-4"
                >
                  <div
                    aria-label="progressBar"
                    aria-valuemax="100"
                    aria-valuemin="0"
                    aria-valuenow="27.9"
                    class="progress-bar"
                    role="progressbar"
                    style="width: 27.9%; background-color: rgb(122, 175, 252);"
                  />
                </div>
                <p>
                  27.9
                  %
                </p>
              </div>
            </div>
          </button>
          <button
            class="mt-2 d-flex flex-row align-items-center justify-content-center"
            data-testid="percentage-item"
            disabled=""
            type="button"
          >
            <div
              class=" position-relative categoryIcon me-3"
            >
              <div
                class="bg-main-blue component-one-third-border-radius"
                style="background-color: rgb(250, 8, 52);"
              >
                <svg
                  fill="none"
                  height="50"
                  viewBox="0 0 51 54"
                  width="50"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M19.0987 34.4061C17.9703 34.5524 17.1675 35.6406 17.3055 36.8366C17.4434 38.0327 18.47 38.8838 19.5984 38.7375C20.7268 38.5912 21.5297 37.503 21.3917 36.3069C21.2537 35.1108 20.2271 34.2598 19.0987 34.4061Z"
                    fill="white"
                  />
                  <path
                    d="M12.2198 28.7036C12.0818 27.5075 12.8847 26.4193 14.013 26.273C15.1414 26.1267 16.168 26.9778 16.306 28.1739C16.444 29.3699 15.6412 30.4581 14.5128 30.6044C13.3844 30.7507 12.3578 29.8996 12.2198 28.7036Z"
                    fill="white"
                  />
                  <path
                    d="M17.0998 17.0805C15.9714 17.2268 15.1685 18.315 15.3065 19.5111C15.4445 20.7072 16.4711 21.5582 17.5995 21.4119C18.7279 21.2656 19.5308 20.1774 19.3928 18.9814C19.2548 17.7853 18.2282 16.9342 17.0998 17.0805Z"
                    fill="white"
                  />
                  <path
                    d="M33.1102 21.0498C32.9722 19.8537 33.7751 18.7655 34.9035 18.6192C36.0319 18.4729 37.0585 19.324 37.1965 20.5201C37.3345 21.7161 36.5316 22.8043 35.4032 22.9506C34.2748 23.0969 33.2482 22.2459 33.1102 21.0498Z"
                    fill="white"
                  />
                  <path
                    d="M25.4848 24.5353C24.3564 24.6816 23.5535 25.7698 23.6915 26.9659C23.8295 28.1619 24.8561 29.013 25.9845 28.8667C27.1129 28.7204 27.9157 27.6322 27.7777 26.4361C27.6397 25.2401 26.6131 24.389 25.4848 24.5353Z"
                    fill="white"
                  />
                  <path
                    d="M24.2507 16.1535C24.1127 14.9574 24.9156 13.8692 26.0439 13.7229C27.1723 13.5767 28.1989 14.4277 28.3369 15.6238C28.4749 16.8199 27.6721 17.9081 26.5437 18.0543C25.4153 18.2006 24.3887 17.3496 24.2507 16.1535Z"
                    fill="white"
                  />
                  <path
                    clip-rule="evenodd"
                    d="M5.43716 26.4664C5.70577 14.7212 14.906 5.43059 25.9864 5.71531C37.0668 6.00003 45.8315 15.7522 45.5629 27.4975C45.5459 28.2413 45.31 28.9209 44.8754 29.4804C44.4598 30.0156 43.9217 30.3665 43.3907 30.6014C42.3574 31.0586 41.0622 31.2021 39.7805 31.2112C34.2463 31.2509 29.4394 36.2967 29.1252 42.2476C29.0519 43.636 28.8285 45.0482 28.2649 46.1578C27.973 46.7323 27.5631 47.279 26.9888 47.6747C26.401 48.0798 25.7268 48.2669 25.0137 48.2486C13.9333 47.9638 5.16856 38.2116 5.43716 26.4664ZM25.9115 8.9871C16.5358 8.74618 8.75104 16.6074 8.52376 26.5457C8.29648 36.484 15.7128 44.7359 25.0885 44.9768C25.2287 44.9804 25.2805 44.9484 25.3059 44.9309C25.345 44.9039 25.4357 44.8242 25.5464 44.6062C25.788 44.1307 25.9777 43.2927 26.0425 42.0648C26.4428 34.4832 32.5075 27.9906 39.7596 27.9386C40.9165 27.9303 41.7284 27.7905 42.2013 27.5812C42.3683 27.5074 42.4464 27.4449 42.4763 27.4165C42.7027 17.4789 35.2868 9.228 25.9115 8.9871Z"
                    fill="white"
                    fill-rule="evenodd"
                  />
                </svg>
              </div>
            </div>
            <div
              class="d-flex flex-column w-100"
            >
              <div
                class="d-flex justify-content-between"
              >
                <p
                  class="fs-4 mb-0"
                >
                  Groceries
                </p>
                <p
                  class="fs-4 mb-0 fw-medium"
                >
                  1337
                </p>
              </div>
              <div
                class="d-flex w-100 justify-content-between"
              >
                <div
                  class="progress w-100 mt-1 me-4"
                >
                  <div
                    aria-label="progressBar"
                    aria-valuemax="100"
                    aria-valuemin="0"
                    aria-valuenow="22.0"
                    class="progress-bar"
                    role="progressbar"
                    style="width: 22.0%; background-color: rgb(250, 8, 52);"
                  />
                </div>
                <p>
                  22.0
                  %
                </p>
              </div>
            </div>
          </button>
          <button
            class="mt-2 d-flex flex-row align-items-center justify-content-center"
            data-testid="percentage-item"
            type="button"
          >
            <div
              class=" position-relative categoryIcon me-3"
            >
              <svg>
                ShowMore.svg
              </svg>
            </div>
            <div
              class="d-flex flex-column w-100"
            >
              <div
                class="d-flex justify-content-between"
              >
                <p
                  class="fs-4 mb-0"
                >
                  Show more
                </p>
                <p
                  class="fs-4 mb-0 fw-medium"
                >
                  3047
                </p>
              </div>
              <div
                class="d-flex w-100 justify-content-between"
              >
                <div
                  class="progress w-100 mt-1 me-4"
                >
                  <div
                    aria-label="progressBar"
                    aria-valuemax="100"
                    aria-valuemin="0"
                    aria-valuenow="50.1"
                    class="progress-bar"
                    role="progressbar"
                    style="width: 50.1%; background-color: rgb(122, 175, 252);"
                  />
                </div>
                <p>
                  50.1
                  %
                </p>
              </div>
            </div>
          </button>
        </div>
      </div>
    </div>
    <div
      class="statistics col-md-4"
    >
      <div
        class="statistics-menu menu mb-3"
      >
        <div
          class="menu-container d-flex flex-column "
        >
          <h3
            class="fw-medium"
          >
            Statistics
          </h3>
          <p>
            Daily Average: 
            10.00
          </p>
          <p>
            Weekly Average: 
            30.00
          </p>
          <p>
            Monthly Average: 
            30.00
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
